crates/sui-core/src/execution_engine.rs

```diff
@@ -373,70 +406,91 @@ fn pay<S>(
-            } else {
-                // need to take all of this coin and some from a subsequent coin
-                coin.balance.withdraw(coin_value).unwrap();
-                remaining_amount -= coin_value;
+    // walk through the coins from left to right, updating the balance to be the corresponding amount,
+    // with remaining_coin tracking extra coin left or debt to pay.
+    // remaining_coin can temp. go negative, but guaranteed to be non-negative at the end of the function.
+    let mut remaining_coin: i128 = 0;
+    for (cur_coin_idx, (recipient, mut amount)) in recipients.iter().zip(amounts).enumerate() {
+        if cur_coin_idx < coins.len() {
+            // add extra amount to the first coin, so that PaySui tx can use it for gas payment
+            if let Some(gas_budget) = gas_budget {
+                if cur_coin_idx == 0 {
+                    amount += gas_budget;
+                }
             }
+            let coin = &mut coins[cur_coin_idx];
+            remaining_coin += i128::from(coin.value()) - i128::from(amount);
+            coin.balance = Balance::new(amount);
+        } else {
+            let new_coin = Object::new_move(
+                MoveObject::new_coin(
+                    coin_type.clone(),
+                    OBJECT_START_VERSION,
+                    bcs::to_bytes(&Coin::new(UID::new(tx_ctx.fresh_id()), amount))
+                        .expect("Serializing coin value cannot fail"),
+                ),
+                Owner::AddressOwner(*recipient),
+                tx_ctx.digest(),
+            );
+            temporary_store.write_object(new_coin, WriteKind::Create);
+            remaining_coin -= i128::from(amount);
         }
     }
 
-    #[cfg(debug_assertions)]
-    {
-        // double check that we didn't create or destroy money
-        let new_total_coins = coins.iter().fold(0, |acc, c| acc + c.value());
-        assert_eq!(total_coins - new_total_coins, total_amount)
+    // pay off debts using the remaining coins in the input coins
+    for coin in coins.iter_mut().skip(recipients.len()) {
+        if remaining_coin >= 0 {
+            break;
+        }
+        remaining_coin += i128::from(coin.value());
+        coin.balance.withdraw(coin.value())?;
     }
 
-    // update the input coins to reflect the decrease in value.
+    // for input coins in 0..recipients.len(), transfer to recipients based on updated coins.
+    // update the rest of input coins to reflect the decrease in value.
     // if the input coin has value 0, delete it
     for (coin_idx, mut coin_object) in coin_objects.into_iter().enumerate() {
-        let coin = &coins[coin_idx];
-        if coin.value() == 0 {
+        let coin = &mut coins[coin_idx];
+        if coin_idx < recipients.len() {
+            let recipient = recipients[coin_idx];
+            // do not bump version for the first coin, as it will be used as gas payment object,
+            // thus its version will be bumped upon gas deduction.
```

Comment:
we probably do not have to bump version or not after dynamic child change, having this block to be consistent with the transfer version bump logic.